2-collectors/scala-stream-collector/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/metrics/MetricsService.scala

```diff
@@ -0,0 +1,68 @@
+package com.snowplowanalytics.snowplow.collectors.scalastream.metrics
+
+import java.time.Duration
+
+import akka.http.scaladsl.model.{HttpMethod, StatusCode, Uri}
+import com.snowplowanalytics.snowplow.collectors.scalastream.metrics.PrometheusMetricsService.Metrics._
+import com.snowplowanalytics.snowplow.collectors.scalastream.metrics.PrometheusMetricsService.NanosecondsInSecond
+import com.snowplowanalytics.snowplow.collectors.scalastream.model.MetricsConfig
+import io.prometheus.client.exporter.common.TextFormat
+import io.prometheus.client.{CollectorRegistry, Counter, Gauge, Histogram}
+import org.apache.commons.io.output.StringBuilderWriter
+
+trait MetricsService {
+
+  def observeRequest(method: HttpMethod, uri: Uri, status: StatusCode, duration: Duration): Unit
+
+  def report(): String
+
+}
+
+class PrometheusMetricsService(metricsConfig: MetricsConfig) extends MetricsService {
+
+  private val registry = new CollectorRegistry
+  private val requestCounter: Counter =
+    Counter.build(HttpRequestCount, HttpRequestCountHelp).labelNames(Labels: _*).register(registry)
+  private val requestDurationHistogramBuilder = Histogram.build(HttpRequestDuration, HttpRequestDurationHelp).labelNames(Labels: _*)
+  private val requestDurationHistogram: Histogram =
+    metricsConfig.durationBuckets
+      .map(buckets => requestDurationHistogramBuilder.buckets(buckets: _*).register(registry))
+      .getOrElse(requestDurationHistogramBuilder.register(registry))
+
+  private val applicationVersionGauge = Gauge.build("service_version", "Java, scala versions and runnable name")
+    .labelNames("java_version", "scala_version", "runnable")
+    .register(registry)
+
+  applicationVersionGauge.labels(StaticMetrics.javaVersion, StaticMetrics.scalaVersion, StaticMetrics.runnableName).set(1)
+
+  override def observeRequest(method: HttpMethod, uri: Uri, status: StatusCode, duration: Duration): Unit = {
+    val path = uri.path.toString
+    val methodValue = method.value
+    val code = status.intValue().toString
+    requestDurationHistogram.labels(path, methodValue, code).observe(duration.toNanos / NanosecondsInSecond)
+    requestCounter.labels(path, methodValue, code).inc()
+  }
+
+  override def report(): String = {
+    val writer = new StringBuilderWriter()
+    TextFormat.write004(writer, registry.metricFamilySamples())
+    writer.getBuilder.toString
+  }
+
+}
+
+object PrometheusMetricsService {
+
+  final val NanosecondsInSecond: Double = Math.pow(10, 9)
+
+  object Metrics {
+    final val HttpRequestDuration = "http_request_duration_seconds"
+    final val HttpRequestDurationHelp = "Latency per endpoint"
+
+    final val HttpRequestCount = "http_requests_total"
+    final val HttpRequestCountHelp = "Total count of requests to http endpoint"
+
+    final val Labels = Seq("endpoint", "method", "code")
```

Comment:
no need for `final` in objects :+1:

Reply:
Actually constant values are better to be declared using `final` modifier. 
https://www.scala-lang.org/files/archive/spec/2.12/05-classes-and-objects.html
`Note, however, that constant value definitions do require an explicit final modifier, even if they are defined in a final class or object.`

The final variables are set during compile time, while non-final members of object are just method calls.
Consider such example
```
object Sample {
  def main(args: Array[String]): Unit = {
    println(Constants.str1)
    println(Constants.str2)
  }
}

object Constants {
  val str1 = "String1"
  final val str2 = "String2"
}
```
Then compile `scalac -print Sample.scala`

And you can see the difference
```
[[syntax trees at end of                   cleanup]] // Sample.scala
package <empty> {
  object Sample extends Object {
    def main(args: Array[String]): Unit = {
      scala.this.Predef.println(Constants.str1());
      scala.this.Predef.println("String2")
    };
    def <init>(): Sample.type = {
      Sample.super.<init>();
      ()
    }
  };
  object Constants extends Object {
    private[this] val str1: String = _;
    <stable> <accessor> def str1(): String = Constants.this.str1;
    final <stable> <accessor> def str2(): String("String2") = "String2";
    def <init>(): Constants.type = {
      Constants.super.<init>();
      Constants.this.str1 = "String1";
      ()
    }
  }
}
```